File: backend/services/legacy-api/internal/domain/transaction.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type TransactionType string

const (
	TransactionIncome    TransactionType = "income"
	TransactionExpense   TransactionType = "expense"
	TransactionTransfer  TransactionType = "transfer"
	TransactionAdjustment TransactionType = "adjustment"
)

type Transaction struct {
	ID              uuid.UUID
	UserID          uuid.UUID
	AccountID       uuid.UUID
	CategoryID      *uuid.UUID
	CounterpartyAccountID *uuid.UUID // transfers
	AmountMinor     int64
	Currency        string // ISO 4217
	Type            TransactionType
	Description     string
	OccurredAt      time.Time
	Version         int64
	CreatedAt       time.Time
	UpdatedAt       time.Time
	DeletedAt       *time.Time
}

File: mobile/velune/lib/domain/models/transaction.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'transaction.freezed.dart';
part 'transaction.g.dart';

enum TransactionType {
  @JsonValue('income')
  income,
  @JsonValue('expense')
  expense,
  @JsonValue('transfer')
  transfer,
  @JsonValue('adjustment')
  adjustment,
}

@freezed
class Transaction with _$Transaction {
  const factory Transaction({
    required String id,
    required String userId,
    required String accountId,
    String? categoryId,
    String? counterpartyAccountId,
    required int amountMinor,
    required String currency,
    required TransactionType type,
    required String description,
    required String occurredAt,
    required int version,
    required String createdAt,
    required String updatedAt,
    String? deletedAt,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) =>
      _$TransactionFromJson(json);
}

